Add a docker-compose stack (#2)
openfun/nextcloud-shibboleth-sync: README.md

# nextcloud-shibboleth-sync

> This repo contains an app to synchronize a list of identity providers with the 

Reviewer: ```suggestion
This repository contains an app to synchronize a list of identity providers with the 
```

[user_saml](https://github.com/nextcloud/user_saml) Nextcloud application.

> This app is written in pure php without dependency and connect to a postgresql database.

Reviewer: ```suggestion
This app is written in pure PHP with no other dependency; it supposes that your NextCloud instance is connected to a PostgreSQL database.
```

## How to build the image

> The docker image will copy the `app` directory and the `CMD` instruction directly executes the script.

Reviewer: You should maybe say that the base image is our home made one? And add a link to it?

Author: No it's the `php` image.

To build it, run:

```bash

> $ docker build -t fundocker/nextcloud-shibboleth-sync:{TAG} .

Reviewer: `TAG` is an example tag or an environment variable?

Reviewer: Also: why not using `make build`?

Author: Because the `make build` build also the `nextcloud`. It is here only to drive the development task

Reviewer: Maybe this Makefile target should be split into two sub-tasks? You did not answer my first question about what `{TAG}` means to you.

Author: sorry, `{TAG}` should be replaced with a real tag, I choose a template like syntax, maybe you have something better to suggest ?

Reviewer: the thing is that having `{}` suggests a variable to me, why not something like:

```suggestion
$ docker build -t fundocker/nextcloud-shibboleth-sync:my_tag .
```

Author: without `$` ? 

![](https://media.giphy.com/media/l1KsOGE23suZQeCfm/giphy.gif)

Author: I made the change.

Reviewer: This was confusing to me :shrug:

```

## How to use the image

To run the image, several environment variables should be used:

- `NEXTCLOUD_IDPS_METADATA_URL`: the XML url containing the identity providers list
- `SAML_DISPLAYNAME_MAPPING`: Attribute to map the displayname to
- `SAML_EMAIL_MAPPING`: Attribute to map the email address to
- `SAML_GENERAL_UID_MAPPING`: Attribute to map the UID to
- `SAML_GROUP_MAPPING`: Attribute to map the users groups to
- `SAML_SERVICE_PROVIDER_PK`: path to the service provider private key
- `SAML_SERVICE_PROVIDER_X509_CERT`: path to the service provider X509 certificate
- `POSTGRES_DB`: postgresql database name
- `POSTGRES_HOST`: postgresql database host
- `POSTGRES_PASSWORD`: postgresql database password
- `POSTGRES_PORT`: postgresql database port
- `POSTGRES_USER`: postgresql database username

A volume is also used to add the X509 certificate and the private key provided by the identity provider,

> this volume must be mounted in the `/private` directory and the path to each file added to the environment 

Reviewer: Relative path to `/private` or absolute path?

Author: the absolute path `/private`

Reviewer: So maybe it's worth to be mentioned explicitly?

Author: starting a path with a `/` is not explicit ? I though it was a unix-like convention

Reviewer: I am talking about the values of the two environment variables not the volume. Your expression may suggest that environment variable values are relative to the volume mount path.

Author: ok done

variables `SAML_SERVICE_PROVIDER_PK` and `SAML_SERVICE_PROVIDER_X509_CERT`.

> A complete exemple is available in the `docker-compose.yml` file.

Reviewer: ```suggestion
A complete example is available in the [`docker-compose.yml`](./docker-compose.yml) file.
```

## Development workflow

To ease the development, a docker-compose configuration is added. To use this docker-compose environment
a `Makefile` is here to execute all needed operations.

> First you need to bootstrap the project

Reviewer: ```suggestion
First you need to bootstrap the project:
```

```bash
$ make bootstrap
```

This task will build all needed images and then install Nextcloud.

> Once the project bootstrap you can run the `sync` task

Reviewer: ```suggestion
Once the project has been bootstrapped you can run the `sync` task
```

```
$ make sync
```

> If you want to run Nextcloud, you have to execute the `run` task and Nextcloud will be available at address: http://localhost:9000

Reviewer: ```suggestion
If you want to run Nextcloud, you have to execute the `run` task and Nextcloud will be available at the following address: http://localhost:9000
```
